ambry-cloud/src/main/java/com/github/ambry/cloud/CloudStorageCompactor.java

```diff
@@ -103,47 +107,84 @@ public int compactPartitions() {
+      while (compactionInProgress > 0) {
+        try {
+          totalBlobsPurged += executorCompletionService.take().get();
+          compactedPartitionCount++;
+        } catch (ExecutionException ex) {
+          vcrMetrics.compactionFailureCount.inc();
+        }
+        compactionInProgress--;
+      }
+      doneLatch.get().countDown();
+    } catch (Throwable th) {
+      logger.error("Hit exception running compaction task", th);
+    } finally {
+      long compactionTime = (System.currentTimeMillis() - compactionStartTime) / 1000;
+      logger.info("Purged {} blobs in {} partitions taking {} seconds", totalBlobsPurged, compactedPartitionCount,
+          compactionTime);
+    }
+    return totalBlobsPurged;
+  }
+
+  /**
+   * Purge the inactive blobs in the specified partitions.
+   * @param partition the {@link PartitionId} to compact.
+   * @return the total number of blobs purged in the partition.
+   */
+  private int compactPartition(PartitionId partition) throws CloudStorageException {
+    if (isShutDown()) {
       logger.info("Skipping compaction due to shut down.");
       return 0;
     }
 
-    // TODO: adjust count when compaction uses multiple threads
-    doneLatch.set(new CountDownLatch(1));
-
-    Set<PartitionId> partitionsSnapshot = new HashSet<>(partitions);
-    logger.info("Beginning dead blob compaction for {} partitions", partitions.size());
-    long now = System.currentTimeMillis();
-    long compactionStartTime = now;
-    long timeToQuit = now + compactionTimeLimitMs;
-    int totalBlobsPurged = 0;
-    for (PartitionId partitionId : partitionsSnapshot) {
-      String partitionPath = partitionId.toPathString();
-      if (!partitions.contains(partitionId)) {
-        // Looks like partition was reassigned since the loop started, so skip it
-        continue;
-      }
+    logger.info("Beginning dead blob compaction for partition {}", partition);
 
-      try {
-        totalBlobsPurged += cloudDestination.compactPartition(partitionPath);
-      } catch (CloudStorageException ex) {
-        logger.error("Compaction failed for partition {}", partitionPath, ex);
-        vcrMetrics.compactionFailureCount.inc();
-      }
+    String partitionPath = partition.toPathString();
+    if (!partitions.contains(partition)) {
+      // Looks like partition was reassigned since the loop started, so skip it
+      logger.warn("Skipping compaction of Partition {} as the partition was reassigned", partition);
+      return 0;
+    }
 
-      if (System.currentTimeMillis() >= timeToQuit) {
-        logger.info("Compaction terminated due to time limit exceeded.");
-        break;
-      }
-      if (isShuttingDown()) {
-        logger.info("Compaction terminated due to shut down.");
-        break;
-      }
+    try {
+      return cloudDestination.compactPartition(partitionPath);
+    } catch (CloudStorageException ex) {
+      logger.error("Compaction failed for partition {}", partitionPath, ex);
+      throw ex;
```

Comment:
I wonder if exception occurs in this method before we call `executorCompletionService.take().get()`, what will happen? Looks like it may not be captured at line 130, line 147. If it is captured by line 153 `catch (Throwable th)`, will the whole batch compaction be terminated? What about some outstanding threads that are already assigned compaction task?

Reply:
If exception would happen in `compactPartition()`, it would be thrown only when `take().get()` is called. It would have been ignored silently if there was no call to `take().get`.

Reply:
I am ok with it as I don't see any issue regarding ignoring exception silently.

Reply:
thanks.. one thing I wanted to note is that this code will actually never ignore the exception silently.

The way `ExecutorCompletionService::take`  works is that if any thread throws exception, it keeps the exception in its state without throwing out. But then when we call `take()` it looks at the state of the thread, and if the thread has returned some value, then it gives out the value with `.get()` call or if the thread has thrown exception, then it throws that exception on the `take()` call.

So in our code, we will never miss any exception, as we make sure to drain all the threads before shutdown.

Reply:
Thanks for clarification. You mentioned "_The way ExecutorCompletionService::take works is that if any thread throws exception, it keeps the exception in its state without throwing out._".
I would like to know if this also applies to `executorCompletionService.submit()` and `executor.execute()` inside `ExecutorCompletionService`. If I understand correctly, once the task is submitted, `executor.execute()` may happen before `ExecutorCompletionService::take` and at that moment, is any occurred exception still held internally (without being thrown out)?  Correct me if I am wrong, thanks.

Reply:
Your understanding is right. Once the task is submitted, `executor.execute()` may happen anytime. However, if the user task that runs inside the `execute()` throws an exception, `execute` doesn't immediately throw the exception. It catches all the exception (including `RunTimeException`) and keeps it in its internal state. When `take().get()` is called, the `Future::get` method retrieves the outcome of the user task, figures out if the outcome of the task is an exception, and either throws the exception, or returns the result (outcome).
So until the `get` method is called no outcome (either exception or result) of user task is reported.

For more reference, do take a look at the `ThreadPoolExecutor::runWorker` and `FutureTask::report` methods of java executor library.

Reply:
Thanks for thorough explanation, this clarifies everything. I am asking this because I had some issues with `ScheduledExecutorService` due to didn't catch exception correctly.